### PR TITLE
Reland [clang][Sema, Lex, Parse] Preprocessor embed in C and C++

### PR DESCRIPTION
**Pull request by Fznamznon**
This commit implements the entirety of the now-accepted [N3017 -Preprocessor Embed](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n3017.htm) and its sister C++ paper [p1967](https://wg21.link/p1967). It implements everything in the specification, and includes an implementation that drastically improves the time it takes to embed data in specific scenarios (the initialization of character type arrays). The mechanisms used to do this are used under the "as-if" rule, and in general when the system cannot detect it is initializing an array object in a variable declaration, will generate EmbedExpr AST node which will be expanded by AST consumers (CodeGen or constant expression evaluators) or expand embed directive as a comma expression.

This reverts commit https://github.com/llvm/llvm-project/commit/682d461d5a231cee54d65910e6341769419a67d7.

---------

Co-authored-by: The Phantom Derpstorm <phdofthehouse@gmail.com>
Co-authored-by: Aaron Ballman <aaron@aaronballman.com>
Co-authored-by: cor3ntin <corentinjabot@gmail.com>
Co-authored-by: H. Vetinari <h.vetinari@gmx.com>

### CODE REVIEW

**Comment by Fznamznon** [comment]
This fixes https://github.com/llvm/llvm-project/pull/68620#issuecomment-2163448739 .
There was also https://github.com/llvm/llvm-project/pull/68620#issuecomment-2163603239 reported, but I'm not able to access proper logs. The link points to sanitizer buildbots so I suppose it might be the same memory leak.
With this patch and sanitizer build only `Clang-Unit :: Lex/./LexTests/67/120` fails, which also fails on main.

**Review by cor3ntin — APPROVED** [review]
The commit to fix the leak LGTM

File: clang/lib/Basic/IdentifierTable.cpp [hunk 1]
```diff
@@ -441,6 +441,7 @@ tok::PPKeywordKind IdentifierInfo::getPPKeywordID() const {
   CASE( 4, 'e', 's', else);
   CASE( 4, 'l', 'n', line);
   CASE( 4, 's', 'c', sccs);
+  CASE(5, 'e', 'b', embed);
```
Comment (NagyDonat) [inline clang/lib/Basic/IdentifierTable.cpp:444]:
```suggestion
  CASE( 5, 'e', 'b', embed);
```
Just drive-by bikeshedding -- consider aligning this like the lines around it.

File: clang/test/Preprocessor/embed_weird.cpp [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+#define FILE_NAME <null_byte.bin>
+#define LIMIT 1
+#define OFFSET 0
+#define EMPTY_SUFFIX suffix()
+
+constexpr unsigned char ch =
+#embed FILE_NAME limit(LIMIT) clang::offset(OFFSET) EMPTY_SUFFIX
+;
+static_assert(ch == 0);
```
[the remark this answers is not in the document]
Comment (Fznamznon) [reply clang/test/Preprocessor/embed_weird.cpp:98]: The first is interesting, cast to float makes #embed considered as a comma expression, so there is actually not enough arguments. Although, I think this is a correct behavior. Otherwise It is not quite clear to me to which data element the cast should apply.

File: clang/lib/Basic/IdentifierTable.cpp (hunk 1 above)
Comment (Fznamznon) [reply clang/lib/Basic/IdentifierTable.cpp]: Ok, done.

File: clang/test/Preprocessor/embed_weird.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (Fznamznon) [reply clang/test/Preprocessor/embed_weird.cpp:100]: Added cases.

File: clang/lib/StaticAnalyzer/Core/ExprEngine.cpp [hunk 3]
```diff
@@ -2422,6 +2422,10 @@ void ExprEngine::Visit(const Stmt *S, ExplodedNode *Pred,
       Bldr.addNodes(Dst);
       break;
     }
+
+    case Stmt::EmbedExprClass:
+      llvm_unreachable("Support for EmbedExpr is not implemented.");
```
[the remark this answers is not in the document]
Comment (Fznamznon) [reply clang/lib/StaticAnalyzer/Core/ExprEngine.cpp]: Used `report_fatal_error` instead.

File: clang/test/Preprocessor/Inputs/null_byte.bin
[the remark this answers is not in the document]
Comment (Fznamznon) [reply clang/test/Preprocessor/Inputs/null_byte.bin:1]: Ok, removed null byte file.

File: clang/test/Preprocessor/embed_weird.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (AaronBallman) [reply clang/test/Preprocessor/embed_weird.cpp:100]: Because embed produces tokens from the preprocessor, the first example should be the same as:
```
void f(float x, char y, char z);
void g() { f((float)
1, 2, 3
);
}
```
which should be accepted (the cast applies to the first argument). You can run into the same with something like:
```

void f(float x, char y, char z);
void g() {
  f(
#embed "three_character_file" prefix((float))
  );
}
```

[the remark this answers is not in the document]
Comment (Fznamznon) [reply clang/test/Preprocessor/embed_weird.cpp:100]: This is one more case, that would be solved by injecting tokens back into the stream. Right now it is quite complex to understand that embed met by `ParseCastExpression` should be expanded in a special way.

[the remark this answers is not in the document]
Comment (cor3ntin) [reply clang/test/Preprocessor/embed_weird.cpp:100]: > This is one more case, that would be solved by injecting tokens back into the stream. Right now it is quite complex to understand that embed met by `ParseCastExpression` should be expanded in a special way.

+1. I can work on that when I come back from st Louis if you don't get to it first

[the remark this answers is not in the document]
Comment (jakubjelinek) [reply clang/test/Preprocessor/embed_weird.cpp:100]: Unless you want to special case it in way too many spots, I'd think it would be far easier to optimize just the inner part of the integer sequence, i.e. everything except the first and last sequence element (maybe with the exception when the last prefix token is , or first suffix token is ,
Because one can use arbitrary tokens before and after the #embed, it can be
```c
const unsigned char a[] = {
-400 + 4 * 
#embed __FILE__
- 27 };
```
(or with tokens from prefix/suffix) and at least the current patchset mishandles many of such cases.  For the inner part of the sequence you know there is , before it and , after it, which simplifies a lot of things.
The above is handled correctly by GCC and by clang -save-temps, but not by clang without -save-temps.
And there are tons of other cases like that, e.g. even designated initializer [26] = 
before the sequence, etc.

[the remark this answers is not in the document]
Comment (Fznamznon) [reply clang/test/Preprocessor/embed_weird.cpp:100]: I have a prototype of injecting tokens that helps. It also removes all the "whack a mole" around template arguments. The only downside it is now yields int instead of unsigned char, but I guess it is fine?

Should I push it to this PR or it makes sense to land this first and make a separate PR? NOTE: I'm on vacation next week, so I will not be available.

[the remark this answers is not in the document]
Comment (AaronBallman) [reply clang/test/Preprocessor/embed_weird.cpp:100]: > I have a prototype of injecting tokens that helps. It also removes all the "whack a mole" around template arguments. The only downside it is now yields int instead of unsigned char, but I guess it is fine?

Nice! Yes, it's fine to yield an `int`; that's how the feature is defined to behave in C and we need the semantics to be the same in C and C++.

> Should I push it to this PR or it makes sense to land this first and make a separate PR? NOTE: I'm on vacation next week, so I will not be available.

IMO, it would be easier for reviewers to land the current changes and then push fixes and improvements separately. This patch is already really hard to review due to size. What do folks think about landing the changes as-is today/tomorrow and then doing follow-up work once @Fznamznon is back from vacation?

**Review by AaronBallman — APPROVED** [review]
Leak fix LGTM, I think it's ready to re-land and try again.

**Comment by Fznamznon** [comment]
Buildbot failure, I'm looking https://lab.llvm.org/buildbot/#/builders/176/builds/226 .

**Comment by Fznamznon** [comment]
@tbaederr , I noticed that all buildbot failures relate to the run with the new constant interpreter. I was wondering if you could see if I did something wrong? For example, embed by default yields values of type `unsigned char`. However when expanding in [ByteCodeExprGen.cpp](https://github.com/llvm/llvm-project/pull/95802/files#diff-ccb62cc00bada13b706286e9cdc64881c86d81da71f2939bcbb40ee692299d67) , I did not insert any casts. I think this could affect, since I had to insert casts in other places and from the log (as @AaronBallman noticed):
```
Line 18: in call to 'value(1778384896, 1795162112)'
The first value is 0x6A000000 in hex and the second is 0x6B000000. The ASCII value for j is 0x6A and k is 0x6B
```
So the values are actually correct

**Comment by tbaederr** [comment]
Do you have a smaller reproducer? Are all the failing build bots big endian?

**Comment by Fznamznon** [comment]
Yes, all bots are big endian. Reproducer is
```
clang -cc1 %s -fsyntax-only -verify -fexperimental-new-constant-interpreter
constexpr int value(int a, int b) {
  return a + b;
}
constexpr int init_list_expr() {
  int vals[] = {
#embed "jk.txt"
  };
  return value(vals[0], vals[1]);
}
constexpr int ExpectedValue = 'j' + 'k';
static_assert(init_list_expr() == ExpectedValue);
```

contents of "jk.txt" is simply "jk".

**Comment by Fznamznon** [comment]
I'm trying to insert a cast using emitCast:
```
--- a/clang/lib/AST/Interp/ByteCodeExprGen.cpp
+++ b/clang/lib/AST/Interp/ByteCodeExprGen.cpp
@@ -1347,6 +1347,13 @@ bool ByteCodeExprGen<Emitter>::visitInitList(ArrayRef<const Expr *> Inits,
     }

     auto Eval = [&](Expr *Init, unsigned ElemIndex) {
+      auto ArrayTy = Ctx.getASTContext().getAsConstantArrayType(T);
+      std::optional<PrimType> FromT = classify(Init->getType());
+      std::optional<PrimType> ToT = classify(ArrayTy->getElementType());
+      if (FromT != ToT) {
+        if (!this->emitCast(*FromT, *ToT, Init))
+          return false;
+      }
       return visitArrayElemInit(ElemIndex, Init);
     };
```
But it fails with an assertion
```
source/llvm-project/clang/lib/AST/Interp/InterpStack.h:46: T clang::interp::InterpStack::pop() [with T = clang::interp::Integral<8, false>]: Asserti
on `ItemTypes.back() == toPrimType<T>()' failed.
```

**Comment by tbaederr** [comment]
Here's a quick patch with the cast inserted: 

```diff
diff --git a/clang/lib/AST/Interp/ByteCodeExprGen.cpp b/clang/lib/AST/Interp/ByteCodeExprGen.cpp
index 731153a6ead9..e7fa1a62c277 100644
--- a/clang/lib/AST/Interp/ByteCodeExprGen.cpp
+++ b/clang/lib/AST/Interp/ByteCodeExprGen.cpp
@@ -1346,13 +1346,30 @@ bool ByteCodeExprGen<Emitter>::visitInitList(ArrayRef<const Expr *> Inits,
       }
     }

-    auto Eval = [&](Expr *Init, unsigned ElemIndex) {
-      return visitArrayElemInit(ElemIndex, Init);
-    };
-
+    E->dump();
     unsigned ElementIndex = 0;
     for (const Expr *Init : Inits) {
-      if (auto *EmbedS = dyn_cast<EmbedExpr>(Init->IgnoreParenImpCasts())) {
+      if (const auto *EmbedS = dyn_cast<EmbedExpr>(Init->IgnoreParenImpCasts())) {
+        QualType TargetType = Init->getType();
+        PrimType TargetT = classifyPrim(Init->getType());
+        TargetType->dump();
+
+
+    auto Eval = [&](const Expr *Init, unsigned ElemIndex) {
+      PrimType InitT = classifyPrim(Init->getType());
+      if (!this->visit(Init))
+        return false;
+      if (InitT != TargetT) {
+        if (!this->emitCast(InitT, TargetT, E))
+          return false;
+      }
+    return this->emitInitElem(TargetT, ElemIndex, Init);
+    };
+
+
+
+
+
         if (!EmbedS->doForEachDataElement(Eval, ElementIndex))
           return false;
       } else {
```

Can you check if that fixes the problem?

**Comment by Fznamznon** [comment]
I can't because I don't have a big endian to verify with. We can try to push speculatively if it doesn't break existing tests.

**Comment by Fznamznon** [comment]
I wonder if that would be ok to disable interpreter tests for now?

**Comment by tbaederr** [comment]
I just pushed https://github.com/llvm/llvm-project/commit/99f5fcb0d1e04125daa404ff14c9cd14b7a2c40b - I don't have time to run _all_ the tests though, so this is a bit of a long shot. If that doesn't  fix it, then disabling them for now sounds fine to me.

**Comment by Fznamznon** [comment]
Thank you!

**Comment by Fznamznon** [comment]
According to the bots that worked!

**Comment by AaronBallman** [comment]
Thank you both for collaborating to get that solved!

**Comment by jakubjelinek** [comment]
@ThePhD @AaronBallman @cor3ntin
Joseph Myers raised an interesting question whether the tokens in embed-parameter-sequence are macro expanded or not.
Consider
```c
#define FILE "/etc/passwd"
#define LIMIT limit(1)
#define THIS , 1, 2, 3
#define PRE prefix (42,
ONE
#embed FILE LIMIT suffix(THIS) PRE )
TWO
#embed "/etc/passwd" LIMIT suffix(THIS) PRE )
THREE
#define limit prefix
#embed "/etc/passwd" limit (4) suffix (THIS)
```

The first #embed is I hope clear in that it is the #embed pp-tokens new-line case where everything gets macro expanded.
The second case is less clear, the filename part matches the " q-char-sequence " case, but LIMIT suffix(THIS) PRE ) is not valid embed-parameter-sequence, so shouldn't it be expanded too?
And the last case by strict reading shouldn't be macro expanded because it is valid embed-parameter-sequence, still
both the the.phd branch and clang trunk and also my GCC patchset handle it as prefix (4) suffix(, 1, 2, 3).
Would in that reading #embed "/etc/passwd" LIMIT suffix(THIS) be valid too and thus not macro expanded and thus later invalid?
And, if embed-parameter-sequence tokens are macro expanded only sometimes, what should happen say with
#define ARG 2) if_empty (1
#embed "file" limit (ARG)
?
On
```c
#define A "/etc/passwd" limit (1) ) + (0
#if __has_embed (A)
int i;
#endif
```
also all 3 compilers agree and happily use the closing ) from the macro to use the ) from the macro expansion for the closing ) of __has_embed and continue through to the rest of the expression.  Or is __has_embed supposed to be always macro expanded?

File: clang/include/clang/Basic/DiagnosticLexKinds.td [hunk 4]
```diff
@@ -436,6 +436,14 @@ def warn_cxx23_compat_warning_directive : Warning<
 def warn_c23_compat_warning_directive : Warning<
   "#warning is incompatible with C standards before C23">,
   InGroup<CPre23Compat>, DefaultIgnore;
+def ext_pp_embed_directive : ExtWarn<
```
Comment (karka228) [inline clang/include/clang/Basic/DiagnosticLexKinds.td:439]: Is there a reason why this is a `ExtWarn` and not `Extension`? The difference as I understand it is that with `ExtWarn` you always get the warning if you compile with a C standard before C23, but with `Extension` you get it only with the option `-Wpedantic`. Isn't that how it is normally handled?

If we compare with `_BitInt` that also is a C23 feature it is handled as `Extension`:
```
def ext_bit_int : Extension<
  "'_BitInt' in %select{C17 and earlier|C++}0 is a Clang extension">,
```

Comment (AaronBallman) [reply clang/include/clang/Basic/DiagnosticLexKinds.td:439]: > Is there a reason why this is a ExtWarn and not Extension? 

We're consistently inconsistent; if you look through `DiagnosticParseKinds.td` you'll see we're all over the place. I think the organic rule tends to be whether the patch author thinks the failure mode from not noticing use of the extension will be disruptive or not.

I think for the preprocessor we'd generally use `ExtWarn` for new directives because those are *really* not portable and can lead to some very odd follow-on diagnostics. That's why we went with `ExtWarn` for `#elifdef` (also new in C23), though I think it's less of a concern with `#embed`.
